site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,622 @@
+
+I'll go through the template line by line:
+
+- **2 - 6**: Define the new struct
+- **7**: Iterate over methods in our metadata
+- **8**: Define a function on the new struct that has exactly the same signature as original one
+- **9**: Print all function parameters by iterating on `$element.Params` using the helper functions defined above
+- **10 - 14**: Run the actual code and either exit the function or return the results, depending on function signature
+
+For the `Timer` decorator, we'll write the following template:
+
+{{< highlight go "linenos=table" >}}
+	timerTemplate := `
+	// Generated code; DO NOT EDIT.
+	package animals
+
+	type AnimalTimer struct {
+		Speaker
+	}
+	{{range $index, $element := .}}
+	func (a *AnimalTimer) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+		fmt.Println("Timing {{$index}} function...")
+		__t := time.Now()
+		{{- if $element.Results | len | eq 0}}
+			a.Speaker.{{$index}}({{$element.Params | joinParams}})
+		{{else}}
+			ret := a.Speaker.{{$index}}({{$element.Params | joinParams}})
+		{{end}}	
+		fmt.Printf("{{$index}} took %s\n", x, time.Since(__t))
+		{{- if not ($element.Results | len | eq 0)}}
+		return ret
+		{{end}}	
+
+	}
+	{{end}}
+	`
+{{< / highlight >}}
+
+This is very similar to the template above, only this time we record the start time of the function and print the elapsed time on exit.
+
+With these templates in hand, we can now generate the decorators!
+
+```go
+	// Create output buffer
+	out := bytes.NewBufferString("")
+	// Parse the template and pass it the helper functions
+	t := template.Must(template.New("my.go.tmpl").Funcs(helperFuncs).Parse(tracerTemplate))
+	// Execute the template and pass it the metadata we collected before
+	t.Execute(out, metadata)
+	// Add needed imports and format the code before printing
+	formattedCode, err := imports.Process("animal_tracer.go", out.Bytes(), &imports.Options{Comments: true})
+	if err != nil {
+		fmt.Printf("cannot format source code, might be an error in template: %s\n", err)
+		return err
+	}
+	// print it out!
+	fmt.Println(string(formattedCode))
+```
+
+The result will be:
+
+```go
+package animals
+
+import "fmt"
+
+type AnimalTracer struct {
+	Speaker
+}
+
+func (a *AnimalTracer) Speak(x int) {
+	fmt.Printf("Running Speak(x) with 'x'=%v ", x)
+	a.Speaker.Speak(x)
+}
+```
+
+Beautiful!
+
+Similarly, running the generator over the `timerTemplate` will yield:
+
+```go
+package animals
+
+import (
+	"fmt"
+	"time"
+)
+
+type AnimalTimer struct {
+	Speaker
+}
+
+func (a *AnimalTimer) Speak(x int) {
+	fmt.Println("Timing Speak function...")
+	__t := time.Now()
+	a.Speaker.Speak(x)
+	fmt.Printf("Speak took %s\n", x, time.Since(__t))
+}
+```
+
+## Finishing up
+
+Using the techniques from the previous section, we can now generate the OpenTracing decorator we want by using the following template:
+
+```go
```

Comment:
To be more consistent, should we follow the same annotation for templates here?

Reply:
thanks @isacikgoz , I'm not sure what you mean in this comment though. Can you elaborate?

Reply:
I meant using annotation like below:
```suggestion
{{< highlight go "linenos=table" >}}
```

Reply:
I only used {{< highlight go "linenos=table" >}} for the example where line numbers were needed. Using it here won't change the syntax highlighting (it doesn't really support templates)

Reply:
I see 👍